File: cpp/src/dsp.cpp
#include "reality_bridge/dsp.hpp"
#include <algorithm>
#include <cmath>
#include <stdexcept>
namespace reality_bridge {
double midi_to_hz(double midi,double a4) noexcept{return a4*std::pow(2.0,(midi-69.0)/12.0);} 
double hz_to_midi(double hz,double a4){if(!(hz>0.0))throw std::invalid_argument("frequency must be positive");return 69.0+12.0*std::log2(hz/a4);} 
KarplusStrongString::KarplusStrongString(double sr){set_sample_rate(sr);} 
void KarplusStrongString::set_sample_rate(double sr){sample_rate_=std::clamp(sr,8000.0,384000.0);stop();}
float KarplusStrongString::noise() noexcept{rng_^=rng_<<13;rng_^=rng_>>17;rng_^=rng_<<5;return(static_cast<float>(rng_&0xFFFFu)/32767.5f)-1.0f;}
void KarplusStrongString::pluck(double hz,double force,double damping,double brightness,double pick_position){const double maxf=sample_rate_*.45;hz=std::clamp(hz,20.0,maxf);force=std::clamp(force,0.0,1.0);damping_=static_cast<float>(std::clamp(damping,.85,.99995));brightness_=static_cast<float>(std::clamp(brightness,.02,.995));const std::size_t n=std::max<std::size_t>(2,static_cast<std::size_t>(std::llround(sample_rate_/hz)));delay_.assign(n,0);index_=0;previous_=0;const float pick=static_cast<float>(std::clamp(pick_position,.02,.98));const float amp=static_cast<float>(force)*.65f;const std::size_t notch=std::max<std::size_t>(1,static_cast<std::size_t>(pick*n));for(std::size_t i=0;i<n;++i){float v=noise()*amp;if(i>=notch)v-=.45f*delay_[i-notch];delay_[i]=v;}envelope_=1;active_=true;}
float KarplusStrongString::process() noexcept{if(!active_||delay_.empty())return 0;const auto next=(index_+1)%delay_.size();const float current=delay_[index_],avg=.5f*(current+delay_[next]),filtered=brightness_*avg+(1-brightness_)*previous_;previous_=filtered;delay_[index_]=filtered*damping_;index_=next;envelope_*=.999985f;if(std::abs(current)<1e-5f&&envelope_<.02f)stop();return current*envelope_;}
void KarplusStrongString::stop() noexcept{active_=false;delay_.clear();index_=0;previous_=0;envelope_=0;}
void SoftLimiter::set_drive(float drive) noexcept{drive_=std::clamp(drive,.1f,8.f);} 
float SoftLimiter::process(float x) const noexcept{return std::tanh(x*drive_);} 
StereoBody::StereoBody(double sr){set_sample_rate(sr);} 
void StereoBody::set_sample_rate(double sr){sr_=std::clamp(sr,8000.0,384000.0);z1_l_=z1_r_=0;}
void StereoBody::set_size(float size) noexcept{size_=std::clamp(size,0.f,1.f);} 
void StereoBody::set_resonance(float r) noexcept{resonance_=std::clamp(r,0.f,.98f);} 
void StereoBody::process(float in,float& left,float& right) noexcept{const float c1=.12f+.30f*size_,c2=.18f+.24f*(1-size_);z1_l_=in*c1+z1_l_*(resonance_*(1-c1));z1_r_=in*c2+z1_r_*(resonance_*(1-c2));left=in*.72f+z1_l_*.46f;right=in*.72f+z1_r_*.46f;}
}

File: rust/src/dsp.rs
use std::f64::consts::LN_2;

pub fn midi_to_hz(midi: f64, a4: f64) -> f64 {
    let ref_hz = if a4.is_finite() && a4 > 0.0 { a4 } else { 440.0 };
    ref_hz * 2.0_f64.powf((midi - 69.0) / 12.0)
}

pub fn hz_to_midi(hz: f64, a4: f64) -> f64 {
    let ref_hz = if a4.is_finite() && a4 > 0.0 { a4 } else { 440.0 };
    if !hz.is_finite() || hz <= 0.0 { return f64::NEG_INFINITY; }
    69.0 + 12.0 * ((hz / ref_hz).ln() / LN_2)
}

pub fn soft_limit(sample: f32, drive: f32) -> f32 {
    let d = if drive.is_finite() { drive.abs().max(0.0001) } else { 1.0 };
    let norm = d.tanh().max(0.0001);
    ((sample * d).tanh() / norm).clamp(-1.0, 1.0)
}

#[derive(Clone, Debug)]
struct XorShift64 { state: u64 }
impl XorShift64 {
    fn new(seed: u64) -> Self { Self { state: seed.max(1) } }
    fn next_f32(&mut self) -> f32 {
        let mut x = self.state;
        x ^= x << 13; x ^= x >> 7; x ^= x << 17; self.state = x;
        let unit = (x as f64 / u64::MAX as f64) as f32;
        unit * 2.0 - 1.0
    }
}

#[derive(Debug)]
pub struct StringVoice {
    sample_rate: f64,
    buffer: Vec<f32>,
    index: usize,
    damping: f32,
    brightness: f32,
    pick_position: f32,
    last: f32,
    active: bool,
    rng: XorShift64,
}

impl StringVoice {
    pub fn new(sample_rate: f64) -> Self {
        Self { sample_rate: sample_rate.max(8_000.0), buffer: vec![0.0; 2], index: 0, damping: 0.996, brightness: 0.5, pick_position: 0.5, last: 0.0, active: false, rng: XorShift64::new(0x434f_534d_4f53_0667) }
    }
    pub fn pluck(&mut self, hz: f64, force: f32, damping: f32, brightness: f32, pick_position: f32) {
        let safe_hz = hz.clamp(20.0, self.sample_rate * 0.45);
        let delay = (self.sample_rate / safe_hz).round().max(2.0) as usize;
        self.buffer.resize(delay, 0.0);
        self.index = 0;
        self.damping = (0.985 + damping.clamp(0.0, 1.0) * 0.0145).clamp(0.0, 0.9998);
        self.brightness = brightness.clamp(0.0, 1.0);
        self.pick_position = pick_position.clamp(0.02, 0.98);
        self.last = 0.0;
        self.active = true;
        let gain = force.clamp(0.0, 1.0);
        let pick = self.pick_position;
        let len = self.buffer.len() as f32;
        for (i, sample) in self.buffer.iter_mut().enumerate() {
            let phase = i as f32 / len;
            let pick_notch = ((phase - pick).abs() * std::f32::consts::PI).sin().abs();
            *sample = self.rng.next_f32() * gain * (0.35 + 0.65 * self.brightness) * pick_notch;
        }
    }
    pub fn process(&mut self) -> f32 {
        if !self.active || self.buffer.len() < 2 { return 0.0; }
        let next = (self.index + 1) % self.buffer.len();
        let a = self.buffer[self.index];
        let b = self.buffer[next];
        let average = (a + b) * 0.5;
        let filtered = average * (1.0 - self.brightness * 0.32) + a * (self.brightness * 0.32);
        let feedback = filtered * self.damping;
        self.buffer[self.index] = feedback;
        self.index = next;
        self.last = feedback;
        if feedback.abs() < 1e-6 && self.buffer.iter().all(|v| v.abs() < 1e-5) { self.active = false; }
        soft_limit(feedback, 1.25)
    }
    pub fn stop(&mut self) { self.active = false; self.buffer.fill(0.0); self.last = 0.0; }
    pub fn is_active(&self) -> bool { self.active }
    pub fn sample_rate(&self) -> f64 { self.sample_rate }
}
